### ISSUE TITLE
Problems with driver creation - Index number of Loop can change / Fortran writer does not copy tags

### ISSUE THREAD

**Issue by hiker** [stfc/PSyclone#1961]
The original problem was discovered during #1883. Originally, when calling `start_expr`, I added the corresponding Reference to the start variable, which is based on the loop index in the tree (e.g. `loop0_start`), to the children list. Next time `start_expr` was called, it would just return the value from the children list.

The problem with this approach is that if transformations are applied later, the index changes, and a newly introduced loop might have the same index (and therefore same variable name for start/stop) as the original loop had initially (the problem seen in #1883 is a compilation test after colouring is applied). As a short-term fix, I've removed adding the nodes to the children list.

Now, when creating the driver, I noticed that I get incorrect start/stop symbols. When the Fortran backend visits the routine, it creates a copy of the symbol table, but this copy does NOT copy the tags. The DynLoop is still in the tree, so when the DynLoop is visited to be converted to a string, it will use the symbol table without tags to find the start/stop symbols ... and will create new symbols since it uses the tags to look it up (and the copied symbol table does not have that tag, so a new symbol is created).

I don't know what's the proper way to fix this. I see a few options:
1. Maybe during `lower_to_language` the DynLoop should be converted to a Loop, meaning the start/stop expr will be evaluated once. This might work assuming that at this stage we still have a symbol table with tags. I checked with gocean, and it appears that GOLoop remain in the tree, i.e. are not converted to Loops. So not sure what the intended way is here.
2. We create a unique number for each loop, which is fixed, even if the loop is moved in the tree. This avoids the problems of inconsistent indices. @arporter  pointed out that a solution for this might be part of #1731. I had a look and didn't see this.
3. We copy the symbol table with tags in FortranWriter's routine_node. This seems to solve the problem for now, but the follow-up problem would be what if we have the same tag in several SymbolTable? Not sure if this would actually work in the existing code, but tags would be a problem :(

~~While 3. seems to work, I then spotted an additional problem: when I query the start/stop expression while visiting the DynLoop node, the code (after #1883) will verify if the `loop0_start` symbol has the right type, but it hasn't: somehow during copy the precision of the integer is lost - I try to lookup an `i_def` integer, but the symbol in the symbol table is now `Scalar<INTEGER, UNDEFINED>`. I will investigate this in more detail, and if required raise an additional ticket.~~
Edit: This was a bug in the driver and can be ignored.

I have tested 3. and this works for my examples.

**Comment by arporter** [comment]
A couple of things:

 1. I'd expect GOLoop and DynLoop to be lowered to Loop (that's what lowering is after all).
 2. "When the Fortran backend visits the routine, it creates a copy of the symbol table, but this copy does NOT copy the tags." that sounds like a bug to me. Does @sergisiso agree?

**Comment by sergisiso** [comment]
>I'd expect GOLoop and DynLoop to be lowered to Loop (that's what lowering is after all).

Lowering is used when the node/tree changes to a different structure or language nodes. When they are just specializations (they just add additional functionality on top) they not necessarily need to be lowered because in fact the already are `isinstance(node, lowered_node_type)`. E.g. GOLoop and DynLoop can effectively be used as generic Loops (are Loops in OOP terms).

> "When the Fortran backend visits the routine, it creates a copy of the symbol table, but this copy does NOT copy the tags." that sounds like a bug to me. Does @sergisiso agree?

yes, if its a deep_copy, it has a loop on it that copies the tags and make them point to the new symbols. If this is not working it could be a bug.

> The problem with this approach is that if transformations are applied later, the index changes, and a newly introduced loop might have the same index (and therefore same variable name for start/stop) as the original loop had initially 

I have similar situations on Nemo and they don't seem to be a problem. If you put the loop index variables in the ancestor(Routine) they new_symbol(root_name="idx", ..) it creates them as idx_0, idx_1, idx_2, ... In order of creation. If you leave the symbols inside the loop, at the backend it should see the nameclashes in the inner scope and rename them to e.g. idx_0_1.
Are any of these steps failing?

**Comment by hiker** [comment]
> > I'd expect GOLoop and DynLoop to be lowered to Loop (that's what lowering is after all).
> 
> Lowering is used when the node/tree changes to a different structure or language nodes. When they are just specializations (they just add additional functionality on top) they not necessarily need to be lowered because in fact the already are `isinstance(node, lowered_node_type)`. E.g. GOLoop and DynLoop can effectively be used as generic Loops (are Loops in OOP terms).

This is consistent with what I see in GOcean (i.e. the GOLoop is still in the PSyIR). Thanks for confirming this.

> 
> > "When the Fortran backend visits the routine, it creates a copy of the symbol table, but this copy does NOT copy the tags." that sounds like a bug to me. Does @sergisiso agree?
> 
> yes, if its a deep_copy, it has a loop on it that copies the tags and make them point to the new symbols. If this is not working it could be a bug.

Maybe bad wording from me. It's not actually a copy, it happens when combining the several levels of symbol table into one. This does not call the `deep_copy` function (which would indeed copy tags), but manually loops over all symbol table and all symbols in each table and adds them to the new table ... but without tags. It's easy to add if required (I kept a copy just in case).

But @arporter and I weren't entirely sure if this is on purpose. I can see a potential problem if two symbol tables contain the same tag, but different symbol names? That would be messy - though I am not sure if this would actually work atm. I _think_ we would need symbol-table specific tags or so (which could be done by giving each symbol table a unique number, and pre-pending this number to each tag) I.e. you would query a symbol table for tag `loop0_start`, but it would internally query for (say) `3|loop0_start`, with 3 being the symbol table index. But that's for a different/new ticket :)

So @arporter and I agreed that I would for now add a `lower_to_language_level` method to `DynLoop`, which would basically fix the start/stop expressions. This way, there is no need to look up the symbols when querying `start_expr` etc during the `visit()`, and it all works. Once you fixed #1731 (which I hope/expect would mean the start/stop expressions will get fixed earlier), this function can simply be removed. I've added a todo to 1731 to this small function. But we can discuss different approaches in the telco.

> 
> > The problem with this approach is that if transformations are applied later, the index changes, and a newly introduced loop might have the same index (and therefore same variable name for start/stop) as the original loop had initially
> 
> I have similar situations on Nemo and they don't seem to be a problem. If you put the loop index variables in the ancestor(Routine) they new_symbol(root_name="idx", ..) it creates them as idx_0, idx_1, idx_2, ... In order of creation. If you leave the symbols inside the loop, at the backend it should see the nameclashes in the inner scope and rename them to e.g. idx_0_1. Are any of these steps failing?

What I see is that each DynLoop has a index (based on its position in the routine). Andy detected the following two test failures:
```
FAILED domain/lfric/transformations/dynamo0p3_transformations_test.py::test_haloex_rc4_colouring[False]
FAILED domain/lfric/transformations/dynamo0p3_transformations_test.py::test_haloex_rc4_colouring[True]
```
They are caused because `gen` is called twice: first before applying colouring loop, once after. In between, the loop indices changes. Since they use the same symbol table, things go wrong (I don't remember the exact details, but it was something like previous loop 0 used tag `loop0_start` now became loop1, and tried to use `loop1_start`. This was defined from the previous loop at position 1, and since one of the loops was a colouring loop, it had an invalid stop value, causing the error. I fixed this by removing the code in start/stop expr. that would add the start/stop expressions into the tree, and instead left it the way it was, i.e. always recomputing them.

**Comment by hiker** [comment]
Domain loops ... aaarghh .... but I got that fixed. But that's really annoying, it's the second or third time that I ran into problems because `self.variable` validates the variable, and in a domain loop the `None` then raises an exception. I think we should either accept 'None' as valid when validating variables, or (probably better) split (Dyn?)Loop into two parts - one DynLoop, and a mixing for halo handling or so? Then we could havea domain kernel with just using the mixin.

**Comment by arporter** [comment]
>  I think we should either accept 'None' as valid when validating variables, or (probably better) split (Dyn?)Loop into two parts - one DynLoop, and a mixing for halo handling or so? 

I think we all agree that the halo-exchange placement should be moved out of the Loop node but I don't know whether anyone has a plan for how to do it?

**Comment by hiker** [comment]
I have no plans, esp. since I am not that familiar with the design. There appears to be three different concepts merged into one:
- domain kernels (which are no loop, but are an empty loop)
- a real loop
- halo exchanges
Likely splitting this into three distinct classes seems to be the sensible approach (well, domain kernel might ... just be a kernel??? Or maybe derived from kernel??), with the halo-exchange-class having loops/kernels as children.

**Comment by arporter** [comment]
The only reason domain kernels have an associated loop is because of the halo-exchange logic. Once that is removed from the Loop then we can do away with the 'empty loop' concept.

**Comment by arporter** [comment]
@hiker, can this be closed now that #1964 is on?

**Comment by hiker** [comment]
Yes, it can, and you already did it as I just now noticed :)  Thanks!
